File: cpp/src/llm/use_cases.cpp
/**
 * AlgoForge — src/llm/use_cases.cpp
 * S2 LLM Layer — four use-case free functions.
 *
 * Each function:
 *   1. Builds a prompt from the canonical template in prompts.cpp
 *   2. Calls provider.chat(...)
 *   3. Parses / returns the structured output
 *
 * Float formatting uses snprintf with "%.4f" (locale-independent, §9).
 * Indicator keys are sorted alphabetically (§9).
 */

#include "llm_internal.hpp"

#include <algorithm>
#include <cstdio>
#include <string>
#include <vector>

namespace algoforge::llm {

/* ── Helper: replace first occurrence of placeholder in template ── */
static void replace_placeholder(std::string& s,
                                 const char* placeholder,
                                 const std::string& value) {
    std::string ph(placeholder);
    auto pos = s.find(ph);
    if (pos != std::string::npos) {
        s.replace(pos, ph.size(), value);
    }
}

/* ── Helper: format double as %.4f ── */
static std::string fmt4(double v) {
    char buf[32];
    snprintf(buf, sizeof(buf), "%.4f", v);
    return buf;
}

/* ── Helper: join a list with commas ── */
static std::string csv(const std::vector<std::string>& v) {
    std::string out;
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) out += ',';
        out += v[i];
    }
    return out;
}

/* =========================================================================
 * §7.1  trade_rationale
 * ========================================================================= */

TradeRationaleOutput trade_rationale(
    LLMProvider&               provider,
    const TradeRationaleInput& inp,
    const std::string&         model,
    float                      temperature,
    int                        max_tokens,
    int                        seed)
{
    /* Build indicators JSON: sort by key, use compact format */
    std::string indicators_json = json::emit_indicators(inp.indicators);

    /* Build patterns CSV in input order */
    std::string patterns_csv_str = csv(inp.patterns);

    /* Expand user template */
    std::string user_content(TRADE_RATIONALE_USER_TMPL);
    replace_placeholder(user_content, "{symbol}",              inp.symbol);
    replace_placeholder(user_content, "{side}",                inp.side);
    replace_placeholder(user_content, "{timeframe}",           inp.timeframe);
    replace_placeholder(user_content, "{entry_price}",         fmt4(inp.entry_price));
    replace_placeholder(user_content, "{stop_loss}",           fmt4(inp.stop_loss));
    replace_placeholder(user_content, "{take_profit}",         fmt4(inp.take_profit));
    replace_placeholder(user_content, "{indicators_json_sorted}", indicators_json);
    replace_placeholder(user_content, "{patterns_csv}",        patterns_csv_str);

    ChatRequest req;
    req.model       = model;
    req.temperature = temperature;
    req.max_tokens  = max_tokens;
    req.seed        = seed;
    req.messages    = {
        {TRADE_RATIONALE_SYSTEM, ""},  /* role/content order — fix below */
    };
    /* Build messages properly */
    req.messages.clear();
    req.messages.push_back({"system", TRADE_RATIONALE_SYSTEM});
    req.messages.push_back({"user",   user_content});

    ChatResponse resp = provider.chat(req);

    TradeRationaleOutput out;
    out.explanation = resp.message.content;
    out.model       = resp.model;
    out.tokens      = resp.completion_tokens;
    return out;
}

/* =========================================================================
 * §7.2  backtest_commentary
 * ========================================================================= */

BacktestCommentaryOutput backtest_commentary(
    LLMProvider&                    provider,
    const BacktestCommentaryInput&  inp,
    const std::string&              model,
    float                           temperature,
    int                             max_tokens,
    int                             seed)
{
    std::string user_content(BACKTEST_COMMENTARY_USER_TMPL);
    replace_placeholder(user_content, "{strategy_name}", inp.strategy_name);
    replace_placeholder(user_content, "{period}",        inp.period);
    replace_placeholder(user_content, "{total_return}",  fmt4(inp.total_return));
    replace_placeholder(user_content, "{sharpe}",        fmt4(inp.sharpe));
    replace_placeholder(user_content, "{sortino}",       fmt4(inp.sortino));
    replace_placeholder(user_content, "{max_drawdown}",  fmt4(inp.max_drawdown));
    replace_placeholder(user_content, "{win_rate}",      fmt4(inp.win_rate));
    replace_placeholder(user_content, "{trade_count}",   std::to_string(inp.trade_count));
    replace_placeholder(user_content, "{best_trade}",    fmt4(inp.best_trade));
    replace_placeholder(user_content, "{worst_trade}",   fmt4(inp.worst_trade));

    ChatRequest req;
    req.model       = model;
    req.temperature = temperature;
    req.max_tokens  = max_tokens;
    req.seed        = seed;
    req.messages    = {
        {"system", BACKTEST_COMMENTARY_SYSTEM},
        {"user",   user_content},
    };

    ChatResponse resp = provider.chat(req);

    BacktestCommentaryOutput out;
    out.commentary = resp.message.content;
    out.model      = resp.model;
    out.tokens     = resp.completion_tokens;
    return out;
}

/* =========================================================================
 * §7.3  news_sentiment
 * ========================================================================= */

NewsSentimentOutput news_sentiment(
    LLMProvider&       provider,
    const std::string& headline,
    const std::string& body,
    const std::string& model,
    float              temperature,
    int                max_tokens,
    int                seed)
{
    std::string user_content(NEWS_SENTIMENT_USER_TMPL);
    replace_placeholder(user_content, "{headline}",     headline);
    replace_placeholder(user_content, "{body_or_empty}", body);

    ChatRequest req;
    req.model       = model;
    req.temperature = temperature;
    req.max_tokens  = max_tokens;
    req.seed        = seed;
    req.messages    = {
        {"system", NEWS_SENTIMENT_SYSTEM},
        {"user",   user_content},
    };

    ChatResponse resp = provider.chat(req);

    /* Extract JSON from response text (§9 balanced-brace scanner) */
    std::string raw;
    try {
        raw = json::extract_first_json_object(resp.message.content);
    } catch (const LLMError&) {
        throw;
    }

    json::JsonValue jv;
    try {
        jv = json::parse(raw);
    } catch (const std::exception& e) {
        throw LLMError(LLMErrorKind::decode, std::string("news_sentiment parse: ") + e.what());
    }

    if (jv.type != json::JsonType::Object) {
        throw LLMError(LLMErrorKind::decode, "news_sentiment: response is not a JSON object");
    }

    auto get_str = [&](const char* k) -> std::string {
        if (!jv.has(k)) throw LLMError(LLMErrorKind::decode,
                                        std::string("news_sentiment: missing key: ") + k);
        return jv.get(k).as_str();
    };
    auto get_dbl = [&](const char* k) -> double {
        if (!jv.has(k)) throw LLMError(LLMErrorKind::decode,
                                        std::string("news_sentiment: missing key: ") + k);
        const auto& v = jv.get(k);
        if (v.type == json::JsonType::Number) return v.as_double();
        /* some models quote the number */
        try { return std::stod(v.as_str()); }
        catch (...) {
            throw LLMError(LLMErrorKind::decode,
                           std::string("news_sentiment: non-numeric value for key: ") + k);
        }
    };

    NewsSentimentOutput out;
    out.sentiment  = get_str("sentiment");
    out.score      = get_dbl("score");
    out.confidence = get_dbl("confidence");
    out.reason     = get_str("reason");
    return out;
}

/* =========================================================================
 * §7.4  strategy_describe
 * ========================================================================= */

StrategyDescribeOutput strategy_describe(
    LLMProvider&                   provider,
    const StrategyDescribeInput&   inp,
    const std::string&             model,
    float                          temperature,
    int                            max_tokens,
    int                            seed)
{
    std::string user_content(STRATEGY_DESCRIBE_USER_TMPL);
    replace_placeholder(user_content, "{name}",            inp.name);
    replace_placeholder(user_content, "{timeframe}",       inp.timeframe);
    replace_placeholder(user_content, "{indicators_csv}",  csv(inp.indicators_used));
    replace_placeholder(user_content, "{entry_rules_csv}", csv(inp.entry_rules));
    replace_placeholder(user_content, "{exit_rules_csv}",  csv(inp.exit_rules));
    replace_placeholder(user_content, "{risk_rules_csv}",  csv(inp.risk_rules));

    ChatRequest req;
    req.model       = model;
    req.temperature = temperature;
    req.max_tokens  = max_tokens;
    req.seed        = seed;
    req.messages    = {
        {"system", STRATEGY_DESCRIBE_SYSTEM},
        {"user",   user_content},
    };

    ChatResponse resp = provider.chat(req);

    StrategyDescribeOutput out;
    out.description = resp.message.content;
    out.model       = resp.model;
    out.tokens      = resp.completion_tokens;
    return out;
}

} /* namespace algoforge::llm */

File: js/src/llm/useCases.js
/**
 * useCases.js — Four use-case functions.
 * All floats in prompts use x.toFixed(4) to match Python's f"{x:.4f}".
 * Indicator dicts are serialized with sorted keys, no spaces.
 */

import { LLMError } from './errors.js';
import {
    TRADE_RATIONALE_SYSTEM,
    TRADE_RATIONALE_USER,
    BACKTEST_COMMENTARY_SYSTEM,
    BACKTEST_COMMENTARY_USER,
    NEWS_SENTIMENT_SYSTEM,
    NEWS_SENTIMENT_USER,
    STRATEGY_DESCRIBE_SYSTEM,
    STRATEGY_DESCRIBE_USER,
} from './prompts.js';

const DEFAULT_MODEL = 'llama3.1:8b';
const DEFAULT_TEMPERATURE = 0.2;
const DEFAULT_MAX_TOKENS = 512;
const DEFAULT_SEED = 42;

// ── JSON extractor (spec §9) ──────────────────────────────────────────────────

/**
 * Extract the first balanced JSON object from a string.
 * Respects quoted strings and \" escapes.
 * @param {string} text
 * @returns {object}
 * @throws {LLMError} kind="decode" if no valid JSON object found
 */
export function extractJson(text) {
    let depth = 0;
    let inString = false;
    let escape = false;
    let start = null;

    for (let i = 0; i < text.length; i++) {
        const ch = text[i];
        if (escape) {
            escape = false;
            continue;
        }
        if (ch === '\\' && inString) {
            escape = true;
            continue;
        }
        if (ch === '"') {
            inString = !inString;
            continue;
        }
        if (inString) continue;
        if (ch === '{') {
            if (depth === 0) start = i;
            depth++;
        } else if (ch === '}') {
            depth--;
            if (depth === 0 && start !== null) {
                const candidate = text.slice(start, i + 1);
                try {
                    return JSON.parse(candidate);
                } catch {
                    // keep scanning
                }
                start = null;
            }
        }
    }
    throw new LLMError('decode', 'no valid JSON object found in response');
}

// ── Indicator serialization ───────────────────────────────────────────────────

/**
 * Serialize indicators dict with sorted keys, no spaces — matches Python's
 * json.dumps(d, sort_keys=True, separators=(',', ':'))
 * @param {Object.<string,number>} indicators
 * @returns {string}
 */
function serializeIndicators(indicators) {
    const sortedKeys = Object.keys(indicators).sort();
    return JSON.stringify(indicators, sortedKeys);
}

// ── Simple template substitution ─────────────────────────────────────────────

/**
 * @param {string} template
 * @param {Object.<string,string>} vars
 * @returns {string}
 */
function renderTemplate(template, vars) {
    return template.replace(/\{([^}]+)\}/g, (_, key) => {
        if (!(key in vars)) throw new Error(`Template key not found: ${key}`);
        return vars[key];
    });
}

// ── Use cases ─────────────────────────────────────────────────────────────────

/**
 * Generate a trade rationale explanation.
 * @param {import('./provider.js').LLMProvider} provider
 * @param {import('./types.js').TradeRationaleInput} inp
 * @param {object} [opts]
 * @param {string} [opts.model]
 * @param {number} [opts.temperature]
 * @param {number|null} [opts.maxTokens]
 * @param {number|null} [opts.seed]
 * @returns {Promise<import('./types.js').TradeRationaleOutput>}
 */
export async function tradeRationale(provider, inp, opts = {}) {
    const model = opts.model ?? DEFAULT_MODEL;
    const temperature = opts.temperature ?? DEFAULT_TEMPERATURE;
    const maxTokens = opts.maxTokens !== undefined ? opts.maxTokens : DEFAULT_MAX_TOKENS;
    const seed = opts.seed !== undefined ? opts.seed : DEFAULT_SEED;

    const indicatorsJson = serializeIndicators(inp.indicators);
    const patternsCsv = inp.patterns.join(',');

    const userContent = renderTemplate(TRADE_RATIONALE_USER, {
        symbol: inp.symbol,
        side: inp.side,
        timeframe: inp.timeframe,
        entry_price: inp.entryPrice.toFixed(4),
        stop_loss: inp.stopLoss.toFixed(4),
        take_profit: inp.takeProfit.toFixed(4),
        indicators_json_sorted: indicatorsJson,
        patterns_csv: patternsCsv,
    });

    const req = {
        model,
        messages: [
            { role: 'system', content: TRADE_RATIONALE_SYSTEM },
            { role: 'user', content: userContent },
        ],
        temperature,
        maxTokens,
        seed,
    };

    const resp = await provider.chat(req);
    return {
        explanation: resp.message.content,
        model: resp.model,
        tokens: resp.completionTokens,
    };
}

/**
 * Generate a backtest commentary narrative.
 * @param {import('./provider.js').LLMProvider} provider
 * @param {import('./types.js').BacktestCommentaryInput} inp
 * @param {object} [opts]
 * @param {string} [opts.model]
 * @param {number} [opts.temperature]
 * @param {number|null} [opts.maxTokens]
 * @param {number|null} [opts.seed]
 * @returns {Promise<import('./types.js').BacktestCommentaryOutput>}
 */
export async function backtestCommentary(provider, inp, opts = {}) {
    const model = opts.model ?? DEFAULT_MODEL;
    const temperature = opts.temperature ?? DEFAULT_TEMPERATURE;
    const maxTokens = opts.maxTokens !== undefined ? opts.maxTokens : DEFAULT_MAX_TOKENS;
    const seed = opts.seed !== undefined ? opts.seed : DEFAULT_SEED;

    const userContent = renderTemplate(BACKTEST_COMMENTARY_USER, {
        strategy_name: inp.strategyName,
        period: inp.period,
        total_return: inp.totalReturn.toFixed(4),
        sharpe: inp.sharpe.toFixed(4),
        sortino: inp.sortino.toFixed(4),
        max_drawdown: inp.maxDrawdown.toFixed(4),
        win_rate: inp.winRate.toFixed(4),
        trade_count: String(inp.tradeCount),
        best_trade: inp.bestTrade.toFixed(4),
        worst_trade: inp.worstTrade.toFixed(4),
    });

    const req = {
        model,
        messages: [
            { role: 'system', content: BACKTEST_COMMENTARY_SYSTEM },
            { role: 'user', content: userContent },
        ],
        temperature,
        maxTokens,
        seed,
    };

    const resp = await provider.chat(req);
    return {
        commentary: resp.message.content,
        model: resp.model,
        tokens: resp.completionTokens,
    };
}

/**
 * Classify financial news sentiment.
 * @param {import('./provider.js').LLMProvider} provider
 * @param {string} headline
 * @param {object} [opts]
 * @param {string} [opts.body]
 * @param {string} [opts.model]
 * @param {number} [opts.temperature]
 * @param {number|null} [opts.maxTokens]
 * @param {number|null} [opts.seed]
 * @returns {Promise<import('./types.js').NewsSentimentOutput>}
 */
export async function newsSentiment(provider, headline, opts = {}) {
    const model = opts.model ?? DEFAULT_MODEL;
    const temperature = opts.temperature ?? DEFAULT_TEMPERATURE;
    const maxTokens = opts.maxTokens !== undefined ? opts.maxTokens : DEFAULT_MAX_TOKENS;
    const seed = opts.seed !== undefined ? opts.seed : DEFAULT_SEED;
    const body = opts.body ?? '';

    const userContent = renderTemplate(NEWS_SENTIMENT_USER, {
        headline,
        body_or_empty: body,
    });

    const req = {
        model,
        messages: [
            { role: 'system', content: NEWS_SENTIMENT_SYSTEM },
            { role: 'user', content: userContent },
        ],
        temperature,
        maxTokens,
        seed,
    };

    const resp = await provider.chat(req);
    const data = extractJson(resp.message.content);
    return {
        sentiment: data.sentiment,
        score: Number(data.score),
        confidence: Number(data.confidence),
        reason: data.reason,
    };
}

/**
 * Generate a plain-English strategy description.
 * @param {import('./provider.js').LLMProvider} provider
 * @param {import('./types.js').StrategyDescribeInput} inp
 * @param {object} [opts]
 * @param {string} [opts.model]
 * @param {number} [opts.temperature]
 * @param {number|null} [opts.maxTokens]
 * @param {number|null} [opts.seed]
 * @returns {Promise<import('./types.js').StrategyDescribeOutput>}
 */
export async function strategyDescribe(provider, inp, opts = {}) {
    const model = opts.model ?? DEFAULT_MODEL;
    const temperature = opts.temperature ?? DEFAULT_TEMPERATURE;
    const maxTokens = opts.maxTokens !== undefined ? opts.maxTokens : DEFAULT_MAX_TOKENS;
    const seed = opts.seed !== undefined ? opts.seed : DEFAULT_SEED;

    const userContent = renderTemplate(STRATEGY_DESCRIBE_USER, {
        name: inp.name,
        timeframe: inp.timeframe,
        indicators_csv: inp.indicatorsUsed.join(','),
        entry_rules_csv: inp.entryRules.join(','),
        exit_rules_csv: inp.exitRules.join(','),
        risk_rules_csv: inp.riskRules.join(','),
    });

    const req = {
        model,
        messages: [
            { role: 'system', content: STRATEGY_DESCRIBE_SYSTEM },
            { role: 'user', content: userContent },
        ],
        temperature,
        maxTokens,
        seed,
    };

    const resp = await provider.chat(req);
    return {
        description: resp.message.content,
        model: resp.model,
        tokens: resp.completionTokens,
    };
}
